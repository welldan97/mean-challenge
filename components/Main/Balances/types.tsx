import { DateTime } from 'luxon';

export interface Balance {
  dateTime: DateTime;
  averageBalance: number;
  transactionsCount: number;
}
