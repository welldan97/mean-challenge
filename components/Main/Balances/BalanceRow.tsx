import { memo } from 'react';
import config from '../../../config';
import { Balance } from './types';

interface BalanceRowProps {
  balance: Balance;
}

const BalanceRow = memo(({ balance }: BalanceRowProps) => (
  <tr>
    <td>
      {new Intl.DateTimeFormat('pt-PT', {
        dateStyle: 'short',
        timeStyle: 'short',
      }).format(balance.dateTime.toJSDate())}
    </td>
    <td>{balance.averageBalance / 1000000000}</td>
    {config.showExtraData && <td>{balance.transactionsCount}</td>}
  </tr>
));

export default BalanceRow;
