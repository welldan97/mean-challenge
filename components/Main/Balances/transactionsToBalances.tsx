import groupBy from 'lodash/fp/groupBy';
import maxBy from 'lodash/fp/maxBy';
import minBy from 'lodash/fp/minBy';
import sortBy from 'lodash/fp/sortBy';
import times from 'lodash/fp/times';
import { DateTime, Interval } from 'luxon';

import config from '../../../config';
import { Transaction, Balance } from '../../../lib/types';

const getTransactionsHourInterval = (transactions: Transaction[]) => {
  const startOfHour = DateTime.fromMillis(
    transactions[0].timestamp * 1000,
  ).startOf('hour');

  return Interval.fromDateTimes(startOfHour, startOfHour.endOf('hour'));
};

const getDurationsPerHour = (transactions: Transaction[]) => {
  const interval = getTransactionsHourInterval(transactions);
  return interval
    .splitAt(...transactions.map(t => DateTime.fromMillis(t.timestamp * 1000)))
    .map(i => i.toDuration().as('milliseconds'));
};

const getBalancesPerHour = (transactions: Transaction[]) => [
  transactions[0].balanceBefore,
  ...transactions.map(t => t.balanceAfter),
];

const accumulatedBalances = (durations: number[], balances: number[]) =>
  durations.reduce((acc, duration, index) => {
    const balance = balances[index];
    return acc + duration * balance;
  }, 0);

const calculateAverageBalance = (transactions: Transaction[]) => {
  const totalDuration = getTransactionsHourInterval(transactions)
    .toDuration()
    .as('milliseconds');
  const durations = getDurationsPerHour(transactions);
  const balances = getBalancesPerHour(transactions);

  return accumulatedBalances(durations, balances) / totalDuration;
};

const balancesGroupedByHour = (transactions: Transaction[]) => {
  const groupedByHour = groupBy<Transaction>(
    t =>
      DateTime.fromMillis(t.timestamp * 1000)
        .startOf('hour')
        .toMillis(),
    sortBy(t => t.timestamp, transactions),
  );
  const hours = times(
    n => DateTime.now().startOf('hour').minus({ hour: n }),
    config.hours,
  );

  return hours.map(h => {
    const transactionsGroup = groupedByHour[h.toMillis().toString()] || [];
    return [h, transactionsGroup] as [DateTime, Transaction[]];
  });
};

const latestBalance = (transactions: Transaction[]) =>
  maxBy(t => t.timestamp, transactions)?.balanceAfter || 0;

const earliestBalance = (transactions: Transaction[]) =>
  minBy(t => t.timestamp, transactions)?.balanceBefore;

const transactionsToBalances = (transactions: Transaction[]) => {
  let lastBalance = latestBalance(transactions);

  return balancesGroupedByHour(transactions).map(
    ([dateTime, transactionsGroup]) => {
      const averageBalance = transactionsGroup.length
        ? calculateAverageBalance(transactionsGroup)
        : lastBalance;

      lastBalance = earliestBalance(transactionsGroup) ?? lastBalance;

      return {
        dateTime,
        averageBalance,
        transactionsCount: transactionsGroup.length,
      };
    },
  ) as Balance[];
};

export default transactionsToBalances;
