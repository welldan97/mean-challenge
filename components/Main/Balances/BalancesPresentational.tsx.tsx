import { memo, useMemo } from 'react';
import config from '../../../config';
import BalanceRow from './BalanceRow';
import { Balance } from './types';

interface Props {
  balances: Balance[];
}

const Balances = memo(({ balances }: Props) => (
  <div className="w-[400px]">
    <table className="table table-zebra w-full mb-16">
      <thead>
        <tr>
          <th>Time</th>
          <th>Average Balance</th>
          {config.showExtraData && <th>Transactions Count</th>}
        </tr>
      </thead>
      <tbody>
        {balances.map(b => (
          <BalanceRow key={b.dateTime.toMillis()} balance={b} />
        ))}
      </tbody>
    </table>
  </div>
));

export default Balances;
