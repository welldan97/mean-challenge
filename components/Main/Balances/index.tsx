import { memo, useMemo } from 'react';
import { Transaction } from '../../../lib/types';
import BalancesPresentational from './BalancesPresentational';
import transactionsToBalances from './transactionsToBalances';

interface Props {
  transactions: Transaction[];
}

const Balances = memo(({ transactions }: Props) => {
  const balances = useMemo(
    () => transactionsToBalances(transactions),
    [transactions],
  );

  return <BalancesPresentational balances={balances} />;
});

export default Balances;
