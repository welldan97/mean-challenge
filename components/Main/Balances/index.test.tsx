import { render } from '@testing-library/react';

import transactions from '../../../lib/test/fixtures/transactions';
import Balances from '.';

describe('Transactions', () => {
  test('render', () => {
    const { container } = render(<Balances transactions={transactions} />);

    expect(container).toMatchSnapshot();
  });
});
