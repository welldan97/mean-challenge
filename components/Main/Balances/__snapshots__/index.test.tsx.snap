// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions render 1`] = `
<div>
  <div
    class="w-[400px]"
  >
    <table
      class="table table-zebra w-full mb-16"
    >
      <thead>
        <tr>
          <th>
            Time
          </th>
          <th>
            Average Balance
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            05/10/22, 20:00
          </td>
          <td>
            1.53897932
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 19:00
          </td>
          <td>
            1.53897932
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 18:00
          </td>
          <td>
            1.53897932
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 17:00
          </td>
          <td>
            1.53897932
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 16:00
          </td>
          <td>
            3.5607685756392824
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 15:00
          </td>
          <td>
            3.597715202567512
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 14:00
          </td>
          <td>
            3.64948216
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 13:00
          </td>
          <td>
            3.6916725356084377
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 12:00
          </td>
          <td>
            4.693386241637246
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 11:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 10:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 09:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 08:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 07:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 06:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 05:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 04:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 03:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 02:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 01:00
          </td>
          <td>
            4.70949716
          </td>
        </tr>
        <tr>
          <td>
            05/10/22, 00:00
          </td>
          <td>
            4.61505268932098
          </td>
        </tr>
        <tr>
          <td>
            04/10/22, 23:00
          </td>
          <td>
            2.70949716
          </td>
        </tr>
        <tr>
          <td>
            04/10/22, 22:00
          </td>
          <td>
            2.70949716
          </td>
        </tr>
        <tr>
          <td>
            04/10/22, 21:00
          </td>
          <td>
            2.70949716
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
